postgres.go

```diff
@@ -0,0 +1,374 @@
+package eventmaster
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/lib/pq"
+	_ "github.com/lib/pq"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	eventmaster "github.com/wish/eventmaster/proto"
+)
+
+// implements datastore
+type PostgresStore struct {
+	db *sql.DB
+}
+
+type PostgresConfig struct {
+	Addr        string `json:"addr"`
+	Port        int    `json:"port"`
+	Database    string `json:"database"`
+	ServiceName string `json:"service_name"`
+	Username    string `json:"username"`
+	Password    string `json:"password"`
+}
+
+func NewPostgresStore(c PostgresConfig) (*PostgresStore, error) {
+	var host string
+	if c.ServiceName != "" {
+		host = c.ServiceName + ".service.consul"
+	} else {
+		host = c.Addr
+	}
+	log.Infof("Connecting to postgres: %v", host)
+	psqlInfo := fmt.Sprintf("host=%s port=%d user=%s "+
+		"password=%s dbname=%s sslmode=disable",
+		host, c.Port, c.Username, c.Password, c.Database)
+	db, err := sql.Open("postgres", psqlInfo)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating postgres session")
+	}
+	err = db.Ping()
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating postgres session")
+	}
+	log.Infof("Successfully connected to postgres %s", host)
+
+	return &PostgresStore{
+		db: db,
+	}, nil
+}
+
+func (p *PostgresStore) AddEvent(e *Event) error {
+	data := "{}"
+	if e.Data != nil {
+		dataBytes, err := json.Marshal(e.Data)
+		if err != nil {
+			return err
+		}
+		data = string(dataBytes)
+	}
+
+	tx, err := p.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.Exec("INSERT INTO event "+
+		"(event_id, parent_event_id, dc_id, topic_id, host, target_host_set, \"user\", event_time, tag_set, received_time)"+
+		" VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)",
+		e.EventID, e.ParentEventID, e.DCID, e.TopicID, e.Host, pq.Array(e.TargetHosts), e.User, e.EventTime, pq.Array(e.Tags), e.ReceivedTime)
+
+	if err != nil {
+		tx.Rollback()
```

Comment:
we should error handling rollback failure too. at least log it.

Reply:
Done